src/linear_algebra/tensor_algebra/to_tensor_power.lean

```diff
@@ -0,0 +1,178 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+import linear_algebra.tensor_algebra.basic
+import linear_algebra.tensor_power
+/-!
+# Tensor algebras as direct sums of tensor powers
+
+In this file we show that `tensor_algebra R M` is isomorphic to a direct sum of tensor powers, as
+`tensor_algebra.equiv_direct_sum`.
+-/
+open_locale direct_sum tensor_product
+
+variables {R M : Type*} [comm_semiring R] [add_comm_monoid M] [module R M]
+
+namespace tensor_power
+
+/-- The canonical embedding from a tensor power to the tensor algebra -/
+def to_tensor_algebra {n} : ⨂[R]^n M →ₗ[R] tensor_algebra R M :=
+pi_tensor_product.lift (tensor_algebra.tprod R M n)
+
+@[simp]
+lemma to_tensor_algebra_tprod {n} (x : fin n → M) :
+  tensor_power.to_tensor_algebra (pi_tensor_product.tprod R x) = tensor_algebra.tprod R M n x :=
+pi_tensor_product.lift.tprod _
+
+@[simp]
+lemma to_tensor_algebra_ghas_one :
+  (@graded_monoid.ghas_one.one _ (λ n, ⨂[R]^n M) _ _).to_tensor_algebra = 1 :=
+tensor_power.to_tensor_algebra_tprod _
+
+@[simp]
+lemma to_tensor_algebra_ghas_mul {i j} (a : ⨂[R]^i M) (b : ⨂[R]^j M) :
+  (@graded_monoid.ghas_mul.mul _ (λ n, ⨂[R]^n M) _ _ _ _ a b).to_tensor_algebra
+    = a.to_tensor_algebra * b.to_tensor_algebra :=
+begin
+  -- change `a` and `b` to `tprod R a` and `tprod R b`
+  rw [tensor_power.ghas_mul_eq_coe_linear_map, ←linear_map.compr₂_apply,
+    ←@linear_map.mul_apply' R, ←linear_map.compl₂_apply, ←linear_map.comp_apply],
+  refine linear_map.congr_fun (linear_map.congr_fun _ a) b,
+  clear a b,
+  ext a b,
+  simp only [linear_map.compr₂_apply, linear_map.mul_apply',
+    linear_map.compl₂_apply, linear_map.comp_apply, linear_map.comp_multilinear_map_apply,
+    pi_tensor_product.lift.tprod, tensor_power.tprod_mul_tprod,
+    tensor_power.to_tensor_algebra_tprod, tensor_algebra.tprod_apply, ←ghas_mul_eq_coe_linear_map],
+  refine eq.trans _ list.prod_append,
+  congr',
+  rw [←list.map_of_fn _ (tensor_algebra.ι R), ←list.map_of_fn _ (tensor_algebra.ι R),
+    ←list.map_of_fn _ (tensor_algebra.ι R), ←list.map_append, list.of_fn_fin_append],
+end
+
+@[simp]
+lemma to_tensor_algebra_galgebra_to_fun (r : R) :
+  (@direct_sum.galgebra.to_fun _ R (λ n, ⨂[R]^n M) _ _ _ _ _ _ _ r).to_tensor_algebra
+    = _root_.algebra_map _ _ r :=
+by rw [tensor_power.galgebra_to_fun_def, tensor_power.algebra_map_eq_smul_one, linear_map.map_smul,
+    tensor_power.to_tensor_algebra_ghas_one, algebra.algebra_map_eq_smul_one]
+
+end tensor_power
+
+namespace tensor_algebra
+
+/-- The canonical map from a direct sum of tensor powers to the tensor algebra. -/
+def of_direct_sum : (⨁ n, ⨂[R]^n M) →ₐ[R] tensor_algebra R M :=
```

Comment:
`equiv_direct_sum` currently has simp lemmas that unfold it down the non-equiv versions. That means I don't need to repeat all the helper lemmas I developed for `of_direct_sum` and `to_direct_sum` for the more bundled version.

I think this matches our strategy in general; write lemmas about the least-bundled case, and have the bundled versions unfold when applied.

Reply:
I'd say that strategy is incomplete. Here the more-bundled maps don't take any additional hypotheses, right? So here I would `simp` from the less-bundled spelling to the more-bundled spelling: since there is no case in which the less-bundled spelling is more general, the more-bundled spelling is strictly more useful (so I see the two individual directions as auxiliary definitions instead of worth independently proving results about).

Reply:
> Here the more-bundled maps don't take any additional hypotheses, right?

Correct


> so I see the two individual directions as auxiliary definitions instead of worth independently proving results about

The problem is that I end up needing all the results about them before I can build the equivalence. So then I end up having to repeat all the results.

See [`clifford_algebra.to_even`](https://leanprover-community.github.io/mathlib_docs/find/clifford_algebra.to_even), [`clifford_algebra.of_even`](https://leanprover-community.github.io/mathlib_docs/find/clifford_algebra.of_even), and [`clifford_algebra.equiv_even`](https://leanprover-community.github.io/mathlib_docs/find/clifford_algebra.equiv_even) for a similar example of this pattern.

My general impression when working with these things is that you work with the equiv, apply the equiv lemmas, then unfold to the map and apply the API lemmas; and that this is easy enough to work with.

Reply:
I see, that makes sense.